tdms/src/surface_phasors.cpp

```diff
@@ -0,0 +1,189 @@
+            break;
+          default:
+            throw runtime_error("Dimension was not recognised!");
+            break;
+        }
+        // multiply by phase factors
+        F.multiply_E_by(cphaseTermE);
+        F.multiply_H_by(cphaseTermH);
+
+        // update the master arrays
+        update_surface_EH(frequency_index, vindex, F);
+      }
+    } else {
+#pragma omp for
+      for (vindex = 0; vindex < n_surface_vertices; vindex++) {
+        CellCoordinate current_cell(surface_vertices[0][vindex], surface_vertices[1][vindex],
+                                    surface_vertices[2][vindex]);
+
+        F.Ex = E.xy[current_cell] + E.xz[current_cell];
+        F.Ey = E.yx[current_cell] + E.yz[current_cell];
+        F.Ez = E.zx[current_cell] + E.zy[current_cell];
+        F.Hx = H.xy[current_cell] + H.xz[current_cell];
+        F.Hy = H.yx[current_cell] + H.yz[current_cell];
+        F.Hz = H.zx[current_cell] + H.zy[current_cell];
+        // multiply by phase factors
+        F.multiply_E_by(cphaseTermE);
+        F.multiply_H_by(cphaseTermH);
+
+        // update the master arrays
+        update_surface_EH(frequency_index, vindex, F);
+      }
+    }//end parallel region
+  }
+}
+
+void SurfacePhasors::create_vertex_list(GridLabels input_grid_labels) {
+  // setup vertex list dimensions
+  mwSize dims[2];
+  dims[0] = n_surface_vertices;
+  dims[1] = 3;
+
+  // create vertex list and infer MATLAB data storage location
+  vertex_list = mxCreateNumericArray(2, (const mwSize *) dims, mxDOUBLE_CLASS, mxREAL);
+  vertex_list_data_ptr = cast_matlab_2D_array(mxGetPr((mxArray *) vertex_list), dims[0], dims[1]);
+
+  //now populate the vertex list
+  for (int i = 0; i < n_surface_vertices; i++) {
+    vertex_list_data_ptr[0][i] = input_grid_labels.x[surface_vertices[0][i]];
+    vertex_list_data_ptr[1][i] = input_grid_labels.y[surface_vertices[1][i]];
+    vertex_list_data_ptr[2][i] = input_grid_labels.z[surface_vertices[2][i]];
+  }
+}
+
+void SurfacePhasors::update_surface_EH(int frequency_index, int vertex_index,
+                       FullFieldSnapshot F) {
+  // Ex
+  surface_EHr[frequency_index][0][vertex_index] += real(F.Ex);
+  surface_EHi[frequency_index][0][vertex_index] += imag(F.Ex);
+  // Hx
+  surface_EHr[frequency_index][3][vertex_index] += real(F.Hx);
+  surface_EHi[frequency_index][3][vertex_index] += imag(F.Hx);
+  // Ey
+  surface_EHr[frequency_index][1][vertex_index] += real(F.Ey);
+  surface_EHi[frequency_index][1][vertex_index] += imag(F.Ey);
+  //Hy
+  surface_EHr[frequency_index][4][vertex_index] += real(F.Hy);
+  surface_EHi[frequency_index][4][vertex_index] += imag(F.Hy);
+  //Ez
+  surface_EHr[frequency_index][2][vertex_index] += real(F.Ez);
+  surface_EHi[frequency_index][2][vertex_index] += imag(F.Ez);
+  // Hz
+  surface_EHr[frequency_index][5][vertex_index] += real(F.Hz);
+  surface_EHi[frequency_index][5][vertex_index] += imag(F.Hz);
```

Comment:
Don't have a suggestion for how to improve, but this _must_ be loop-able.
Also: 0, 3, 1, 4, 2, 5 does not respect my understanding of the ordinality of numbers 😛 .